### ISSUE TITLE
Getting page source is failing for one URL

### ISSUE THREAD

**Issue by vigneshrams** [wkeeling/selenium-wire#135]
Here is the code I used,

`urlstring = "https://www.rottentomatoes.com/m/the_hateful_eight"
           _driver.get(urlstring)
            _driver.wait_for_request(urlstring, timeout=30)
            # Get status Code for the URL request made
            first_request = self._driver.requests[0]
            if (first_request.response.status_code == 200):
                htmlRet = _driver.page_source # throws exception`

**Comment by wkeeling** [comment]
Thanks for the issue. I've tried to reproduce using your example code above, and the `driver.page_source` works for me. If I print out `htmlRet` I see the HTML.

Can you post the error and full traceback that you're seeing?

**Comment by vigneshrams** [comment]
Thanks for the reply. It crashes in this line for me.
`htmlRet = _driver.page_source`
I am not sure what I am missing. The weird part is it only happens for the URL I mentioned in the issue.

**Comment by vigneshrams** [comment]
If you don't mind me asking. Can you send me the working code for this URL?
`urlstring = "https://www.rottentomatoes.com/m/the_hateful_eight"`
./Thanks

**Comment by wkeeling** [comment]
Sure, I pretty much used your example code as is:

```
urlstring = "https://www.rottentomatoes.com/m/the_hateful_eight"

driver = webdriver.Firefox()
driver.get(urlstring)
driver.wait_for_request(urlstring, timeout=30)  # Get status Code for the URL request made
first_request = driver.requests[0]

if first_request.response.status_code == 200:
    htmlRet = driver.page_source
    print(htmlRet)
```

The page takes a while to load because there are a large number of assets included by the page.

**Comment by vigneshrams** [comment]
Thanks, I am using a chrome driver. Will it be the reason for the issue? I even tried using the same use-case with selenium but it works there.

**Comment by wkeeling** [comment]
Switching to `webdriver.Chrome()` also seems to work for me. It prints out the HTML source:

```
<!DOCTYPE html><html xmlns="http://www.w3.org/1999/xhtml" lang="en" dir="ltr" xmlns:fb="http://www.facebook.com/2008/fbml" xmlns:og="http://opengraphprotocol.org/schema/"><head prefix="og: http://ogp.me/ns# flixstertomatoes: http://ogp.me/ns/apps/flixstertomatoes#" class="at-element-marker"><script async="" crossorigin="anonymous" type="text/javascript" src="https://ct.contentsquare.net/www/ChangeMonitor-latest.js"></script><script src="https://www.googletagservices.com/activeview/js/current/osd.js?cb=%2Fr20100101"></script><script src="https://connect.facebook.net/en_US/sdk.js?hash=014fb48e3b44a6659882b5634a7bada3&amp;ua=modern_es6" async="" crossorigin="anonymous"></script><script src="https://securepubads.g.doubleclick.net/pagead/js/rum.js"></script><script type="text/javascript" src="https://fandangollc.demdex.net/event?d_nsid=0&amp;d_ld=_ts%3D1595319575380&amp;d_rtbd=json&amp;d_jsonv=1&amp;d_dst=1&amp;d_cb=demdexRequestCallback_0_1595319575380&amp;c_page_uri=www.rottentomatoes.com%2Fm%2Fthe_hateful_eight&amp;c_nbcu_brand=rottentomatoes&amp;c_nbcu_platform=web&amp;c_nbcu-rottentomatoes-contenttype=movie_page&amp;c_nbcu_cleantitle=The%20Hateful%20Eight%202015%20-%20Rotten%20Tomatoes&amp;c_mps_contentid=X665864238832&amp;c_mps_path=%2Fm%2Fthe_hateful_eight&amp;c_mps-rottentomatoes-cats=movie%7Cmovie_page&amp;c_mps-rottentomatoes-cat1=movie&amp;c_mps-rottentomatoes-cat2=movie_page&amp;c_mps_admode=gpt-asynchronous&amp;c_mps_adlazyload=1&amp;c_mps_cag-rottentomatoes-score=75&amp;c_mps_cag-rottentomatoes-certified_fresh=1&amp;c_mps_cag-rottentomatoes-fresh_rotten=fresh&amp;c_mps_cag-rottentomatoes-rating=r&amp;c_mps_cag-rottentomatoes-release=2015&amp;c_mps_cag-rottentomatoes-movieshow=the%20hateful%20eight&amp;c_mps_cag-rottentomatoes-genre=western&amp;c_mps_cag-rottentomatoes-urlid=the_hateful_eight&amp;c_mps_field-rottentomatoes-env=production&amp;c_mps_field-rottentomatoes-rtid=771372896&amp;c_mps_fwssid=rt_movie_moviepage&amp;c_mps_loadset=0&amp;c_mps_adunitid=%2F2620%2Frottentomatoes%2Fmovie%2Fmovie_page&amp;c_pagename=rottentomatoes%7Cmovie_page%7Cmovie%7CX665864238832%7CThe%20Hateful%20Eight%202015%20-%20Rotten%20Tomatoes"></script><script async="" src="//c.conversionlogic.net/js/fandango/c.js"></script><script async="" src="https://honorableland.com/v2xglf05UGcVvzEUFlB6672D6fDwhuTsalYNh9J8Q6zzvZToUzEeznRfzQK-ELINP"></script><script async="" src="//c.amazon-adsystem.com/aax2/apstag.js"></script><script src="https://www.googletagservices.com/tag/js/gpt.js" async="" type="text/javascript"></script><script type="text/javascript" async="" src="https://www.google-analytics.com/plugins/ua/linkid.js"></script><script async="" src="//consent.truste.com/notice?domain=fandango.com&amp;c=teconsent&amp;js=bb&amp;noticeType=bb" id="truste_0.8050813258653984"></script><script async="" src="https://sb.scorecardresearch.com/beacon.js"></script><script async="" src="//www.google-analytics.com/analytics.js"></script><script src="https://connect.facebook.net/signals/config/1147905821962466?v=2.9.22&amp;r=stable" async=""></script><script async="" src="https://connect.facebook.net/en_US/fbevents.js"></script><script async="" src="https://www.googletagmanager.com/gtm.js?id=GTM-M5LKWCR"></script><script src="//mps.nbcuni.com/fetch/ext/load-rottentomatoes-web.js?nowrite=2" id="mps-gpt-shared" class="mps-script"></script>

        <!-- salt=lay-def-02-juRm -->
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <meta http-equiv="x-ua-compatible" content="ie=edge" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        ... snip ...
```

What exception message do you see? Is there a traceback in the log that you can post?

**Comment by vigneshrams** [comment]
Here is the traceback log I got
T08:23:59.927Z	521e3330-f3a2-4ab3-b4c2-d9f6b33b0122	Error making request
Traceback (most recent call last):
  File "/var/task/seleniumwire/proxy/proxy2.py", line 90, in proxy_request
    conn.request(self.command, path, req_body, dict(req.headers))
  File "/var/lang/lib/python3.6/http/client.py", line 1262, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1308, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1257, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1036, in _send_output
    self.send(msg)
  File "/var/lang/lib/python3.6/http/client.py", line 974, in send
    self.connect()
  File "/var/task/seleniumwire/proxy/proxy2.py", line 367, in connect
    super().connect()
  File "/var/lang/lib/python3.6/http/client.py", line 1415, in connect
    super().connect()
  File "/var/lang/lib/python3.6/http/client.py", line 946, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/var/lang/lib/python3.6/socket.py", line 724, in create_connection
    raise err
  File "/var/lang/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
socket.timeout: timed out[ERROR]	2020-07-21T08:24:06.914Z	521e3330-f3a2-4ab3-b4c2-d9f6b33b0122	Error making request
Traceback (most recent call last):
  File "/var/task/seleniumwire/proxy/proxy2.py", line 90, in proxy_request
    conn.request(self.command, path, req_body, dict(req.headers))
  File "/var/lang/lib/python3.6/http/client.py", line 1262, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1308, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1257, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/var/lang/lib/python3.6/http/client.py", line 1036, in _send_output
    self.send(msg)
  File "/var/lang/lib/python3.6/http/client.py", line 974, in send
    self.connect()
  File "/var/task/seleniumwire/proxy/proxy2.py", line 367, in connect
    super().connect()
  File "/var/lang/lib/python3.6/http/client.py", line 1415, in connect
    super().connect()
  File "/var/lang/lib/python3.6/http/client.py", line 946, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/var/lang/lib/python3.6/socket.py", line 724, in create_connection
    raise err
  File "/var/lang/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
socket.timeout: timed out200
Message: unknown error: unknown sessionId
  (Session info: headless chrome=65.0.3325.181)
  (Driver info: chromedriver=2.37.544315 (730aa6a5fdba159ac9f4c1e8cbc59bf1b5ce12b7),platform=Linux 4.14.165-102.205.amzn2.x86_64 x86_64)

**Comment by wkeeling** [comment]
Thanks. Are you using an upstream proxy (do you have a `proxy` configured in your selenium wire options)?

**Comment by vigneshrams** [comment]
Yes, I have Datacenter/Residential Proxy configured with selenium wire. Only this URL fails on every proxy. The remaining sites work well. I don't know What I am missing.

**Comment by wkeeling** [comment]
OK strange. Can you try [switching the backend](https://github.com/wkeeling/selenium-wire#backends) to mitmproxy?

**Comment by vigneshrams** [comment]
Tried to switching to mitmproxy still the same issue. Not sure how to slove it.

**Comment by vigneshrams** [comment]
Any clue on why there is failure? Did you get any info from traceback logs? Please help me on this.

**Comment by wkeeling** [comment]
I guess the issue is possibly due to your datacenter proxy since it is affecting both backends. Can you post the code that you're using, including any proxy configuration?

**Comment by vigneshrams** [comment]
self.getproxy(self.premiumproxy)

        self.chrome_options.add_argument('--headless')
        self.chrome_options.add_argument('--disable-gpu')
        self.chrome_options.add_argument('--window-size=1280x1696')
        self.chrome_options.add_argument('--no-sandbox')
        self.chrome_options.add_argument('--hide-scrollbars')
        #self.chrome_options.add_argument('--enable-logging')
        #self.chrome_options.add_argument('--log-level=0')
        #self.chrome_options.add_argument('--v=99')
        self.chrome_options.add_argument('--single-process')
        self.chrome_options.add_argument('--ignore-certificate-errors')
        self.chrome_options.add_argument('--disable-dev-shm-usage')
        
        # Add header to avoid blocking from the site scraped

        self.chrome_options.add_argument('user-agent=' + self.useragent)

        # disable Image loading
        # chrome_prefs = {}
        # self.chrome_options.experimental_options["prefs"] = chrome_prefs
        # chrome_prefs["profile.default_content_settings"] = {"images": 2}
        # chrome_prefs["profile.managed_default_content_settings"] = {"images": 2}

        self.chrome_options.binary_location = os.getcwd() + "/driver/headless-chromium"
options = {
            'request_storage_base_dir': '/tmp',  # Use /tmp to store captured data
            'verify_ssl': False,  # Don't verify self signed certificate
            'proxy': {
                'http': 'IP_HERE',
                'https': 'IP_HERE',
                'no_proxy': 'localhost,127.0.0.1,dev_server:8080'
            }
        }
        self._driver = webdriver.Chrome(os.getcwd() + "/driver/chromedriver", chrome_options=chromeoptions,
                                        seleniumwire_options=options)

                                        self._driver.get(self.urlstring)
                                            self._driver.wait_for_request(self.urlstring, timeout=30)
                                            time.sleep(1)  

                                            # Get status Code for the URL request made
                                            first_request = self._driver.requests[0]
                                            print(first_request.response.status_code)

                                                htmlRet = self._driver.page_source


For info, I am running this code inside AWS Lambda

**Comment by vigneshrams** [comment]
Am I doing something wrong here? Please help me with it.

**Comment by wkeeling** [comment]
For your proxy config, it's not clear whether you're using the correct syntax. Can you confirm you're using:

```
'proxy': {
    'http': 'http://IP_HERE:PORT',
    'https': 'https://IP_HERE:PORT',
}
```

Notice that you need the `http://` before the IP address string.

**Comment by vigneshrams** [comment]
Hi, Thanks for your help. 
`Message: unknown error: unknown sessionId
  (Session info: headless chrome=65.0.3325.181)
  (Driver info: chromedriver=2.37.544315 (730aa6a5fdba159ac9f4c1e8cbc59bf1b5ce12b7),platform=Linux 4.14.177-104.253.amzn2.x86_64 x86_64)`

I was able to resolve the issue with chromedriver update. I am using Chrome 65. Thanks a lot.

**Comment by malavolti** [comment]
Hi Will,
I don't know if this is the best way to ask you help, sorry.

I have a similar problem with my selenium-wire script and ChromeDriver 73.0.3683.68 on Debian 9 (Python 3.8.3).

This is my code:

```
from seleniumwire import webdriver 
from selenium.common.exceptions import WebDriverException

def getDriver():

    # Configure Web-driver
    chrome_options = webdriver.ChromeOptions()
    chrome_options.headless = True
    chrome_options.page_load_strategy = 'none'
    chrome_options.add_argument('--incognito')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--start-maximized')
    chrome_options.add_argument('--disable-infobars')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--disable-extensions')

    optionssw = {
       'verify_ssl': False,  # Verify SSL certificates but beware of errors with self-signed certificates
       'connection_timeout': 60
    }

    return webdriver.Chrome(PATHCHROMEDRIVER, options=chrome_options, seleniumwire_options=optionssw)

# Create a new instance of the Chrome driver
driver = getDriver()

try:
   urlstring = "https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth"
   driver.get(urlstring)
   driver.wait_for_request(urlstring, timeout=60)  # Get status Code for the URL request made
   first_request = driver.requests[0]
  
   if first_request.response.status_code == 200:
      htmlRet = driver.page_source
      print(htmlRet)

finally:
   driver.quit()
```

But I reach every time the timeout without success.

Can you help me, please?

Thanks a lot Will!

**Comment by wkeeling** [comment]
Hi Marco,
Thanks for your message.
The driver.get() function is usually synchronous, so it shouldn't be
necessary to add driver.wait_for_request() immediately following it.
The call to get() will block until the page has loaded. (I say is
"usually synchronous" because sometimes get() will unblock before every
asset has fully returned but I think that shouldn't be an issue in your
case.)
So I reproduced the issue using your code below, and then removed the
driver.wait_for_request() line, and I was able to see the html source
printed with print(htmlRet)
This does suggest however, that there may be a bug somewhere in
driver.wait_for_request() - as the code you have written ought to have
worked. It looks like wait_for_request() couldn't find the URL - maybe
because of the entityID parameter being itself a URL. I'll perhaps open
a separate issue to track that, but hopefully removing
driver.wait_for_request() fixes your issue in the meantime.
Thanks,Will

**Comment by malavolti** [comment]
Thank you so much Will for the quick answer!
I hope to find a way to load entirely the page provided through some redirection from the urlstring = "https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth" and get its status_code.

Do you have an idea to manage this problem easily?

Thank you very much!
Marco

**Comment by wkeeling** [comment]
Hi Marco,
I see the issue. The URL you pass in entityID ends up being redirected
a couple of times until it reaches the final 200 status - and that's
the page you're looking for?
You may be best to look for something unique in the final URL, and I
notice that URL contains a jessionid parameter. You should be able to
use that in driver.wait_for_request() - for example:
urlstring = "
https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth"driver.get(urlstring)request
 = driver.wait_for_request('jsessionid', timeout=60)  # Get status Code
for the URL request made
print(request.response.status_code)  # Should print 200
if request.response.status_code == 200:    htmlRet =
driver.page_source    print(htmlRet)
Note that driver.wait_for_request() will return the request so you
don't need to look for it in driver.requests.
I've just run the above code and it seems to work.
Thanks,Will

**Comment by malavolti** [comment]
> Hi Marco, I see the issue. The URL you pass in entityID ends up being redirected a couple of times until it reaches the final 200 status - and that's the page you're looking for? 

YES! :)

> You may be best to look for something unique in the final URL, and I notice that URL contains a jessionid parameter. You should be able to use that in driver.wait_for_request() - for example: urlstring = " [https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth"driver.get(urlstring)request](https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth%22driver.get(urlstring)request) = driver.wait_for_request('jsessionid', timeout=60) # Get status Code for the URL request made print(request.response.status_code) # Should print 200 if request.response.status_code == 200: htmlRet = driver.page_source print(htmlRet) Note that driver.wait_for_request() will return the request so you don't need to look for it in driver.requests. I've just run the above code and it seems to work. Thanks,Will

Thank you Will, but the URL loaded after the "...entityID=https..." is not always in the same format unfortunately...

If you have other ideas... I will appreciate them very much!

Kind regards,
Marco

**Comment by wkeeling** [comment]
Hi Marco,
I notice that the jsessionid seems to be independent of the URL passed
in entityID, in that the jsessionid seems to be added dynamically by
the redirect process, or maybe that's not the case? Essentially though
if the goal is to determine the status code of the final request, we'd
need to know something unique about that request in order to identify
it.
There's another slightly uglier alternative, and that is to capture all
requests, and then iterate the requests to find a response with a
content type of 'text/html' - e.g.
urlstring = "
https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://garr-idp-prod.irccs.garr.it/idp/shibboleth"driver.get(urlstring
)
for request in driver.requests:    if request.response and
request.response.headers.get('Content-
Type').startswith('text/html'):        print(request.response.status_co
de)        htmlRet = driver.page_source        print(htmlRet)

If anything else springs to mind, I'll let you know!
Thanks,Will

**Comment by malavolti** [comment]
Hi Will,
could you try this code please on the latest version of selenium-wire? Thank you so much!

```python
#!/usr/bin/env python3

from seleniumwire import webdriver
from selenium.common.exceptions import WebDriverException

def getDriver():
    PATHCHROMEDRIVER = "/PATH/chromedriver"

    # Configure Web-driver
    chrome_options = webdriver.ChromeOptions()
    chrome_options.headless = True
    chrome_options.page_load_strategy = 'none'
    chrome_options.add_argument('--incognito')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--start-maximized')
    chrome_options.add_argument('--disable-infobars')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--disable-extensions')

    optionssw = {
       'verify_ssl': False,
       'connection_timeout': None
    }

    return webdriver.Chrome(PATHCHROMEDRIVER, options=chrome_options, seleniumwire_options=optionssw)

# Create a new instance of the Chrome driver
driver = getDriver()

try:
   #urlstring = "https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://idp.aco.net/idp/shibboleth"
   urlstring = "https://sp24-test.garr.it/Shibboleth.sso/Login?entityID=https://idp.oh-tech.org/idp/shibboleth"
   driver.get(urlstring)

   for request in driver.requests:
       print(request.url)
       if request.response and str(request.response.headers.get('Content-Type')).startswith('text/html'):
          if(request.response.status_code == 302): continue
          print("request.response.status_code")
          htmlRet = driver.page_source
          print("htmlRet")

finally:
   driver.quit()
```

I receive an "SSL: UNKNOWN_PROTOCOL" error instead of the Login page...
Can you help me?

Thank you so much!
Marco

**Comment by wkeeling** [comment]
Hi Marco,

Yes I also get a SSL issue - although it's slightly different:
```
ssl.SSLError: [SSL: PRE_MAC_LENGTH_TOO_LONG] invalid alert (_ssl.c:852)
```
However, if I switch to the [mitmproxy backend](https://github.com/wkeeling/selenium-wire#backends), the code works fine. 

If you want to switch to the mitmproxy backend, install mitmproxy with:

```
pip install mitmproxy
```
and then set the `backend` option in your Selenium Wire options:
```
    optionssw = {
       'backend': 'mitmproxy',
       'verify_ssl': False,
       'connection_timeout': None
    }
```

I'll see if I can figure out why that SSL error is happening with the default backend, but hopefully the mitmproxy backend gets you past the issue in the meantime.
